lib/ansible/modules/storage/purestorage/purefa_facts.py

```diff
@@ -0,0 +1,350 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2018, Simon Dodsley (simon@purestorage.com)
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: purefa_facts
+version_added: '2.6'
+short_description: Collect facts from Pure Storage FlashArray
+description:
+  - Collect facts information from a Pure Storage Flasharray running the
+    Purity//FA operating system. By default, the module will collect basic
+    fact information including hosts, host groups, protection
+    groups and volume counts. Additional fact information can be collected
+    based on the configured set of arguements.
+author:
+  - Simon Dodsley (@sdodsley)
+options:
+  gather_subset:
+    description:
+      - When supplied, this argument will define the facts to be collected.
+        Possible values for this include all, minimum, config, performance,
+        capacity, network, subnet, interfaces, hgroups, pgroups, hosts,
+        volumes and snapshots.
+    required: false
+    default: minimum
+extends_documentation_fragment:
+  - purestorage.fa
+'''
+
+EXAMPLES = r'''
+- name: collect default set of facts
+  purefa_facts:
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+
+- name: collect configuration and capacity facts
+  purefa_facts:
+    gather_subset:
+      - config
+      - capacity
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+
+- name: collect all facts
+  purefa_facts:
+    gather_subset:
+      - all
+    fa_url: 10.10.10.2
+    api_token: e31060a7-21fc-e277-6240-25983c6c4592
+'''
+
+RETURN = r'''
+ansible_facts:
+  description: Returns the facts collected from the FlashArray
+  returned: always
+  type: dict
+'''
+
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pure import get_system, purefa_argument_spec
+
+
+AC_REQUIRED_API_VERSION = ['1.14']
+CAP_REQUIRED_API_VERSION = ['1.6']
+SAN_REQUIRED_API_VERSION = ['1.10']
+
+
+def generate_default_dict(array):
+    default_facts = {}
+    defaults = array.get()
+    api_version = array.get_rest_version()
+    if api_version in AC_REQUIRED_API_VERSION:
+        pods = array.get_pods()
+        default_facts['pods'] = len(pods)
+    hosts = array.list_hosts()
+    snaps = array.list_volumes(snap=True, pending=True)
+    pgroups = array.list_pgroups(pending=True)
+    hgroups = array.list_hgroups()
+    default_facts['array_name'] = defaults['array_name']
+    default_facts['purity_version'] = defaults['version']
+    default_facts['hosts'] = len(hosts)
+    default_facts['snapshots'] = len(snaps)
+    default_facts['protection_groups'] = len(pgroups)
+    default_facts['hostgroups'] = len(hgroups)
```

Comment:
As you say the API is versioned and will always be backwards compatible. It will always return something even if it's only None or 0. There is no fail exception that I'm aware of.

Reply:
@sdodsley sounds good, thank you for the clarification